Track discover

ashiagr opened the pull request:
| 📘 Project: #261 |
|:---:|

Corresponding iOS PR: https://github.com/Automattic/pocket-casts-ios/pull/263

Yet to be added: `discover_shown`

Some of the events are 1:1 implementations of the existing Firebase events, and contain the same properties:
- `discover_list_episode_play`
- `discover_list_podcast_subscribed`
- `discover_list_podcast_tapped`
- `discover_list_episode_tapped`
- `discover_list_show_all_tapped`
- `discover_list_impression`
- `discover_featured_podcast_tapped`

Similar to iOS, I've also added new events:
- `discover_featured_page_changed`: When the user swipes to the next/previous page on the Featured section
- `discover_small_list_page_changed`: When the user swipes to the next/previous page on a "small list" section
    - The name small list comes from the discover json and is used to show items such as Trending and Popular
- iOS tracks below additional events: 
`discover_large_list_page_changed`
`discover_network_list_page_changed`
There's no snap-to or paging behavior for large and network lists in Android,  so they're not tracked.
- `discover_region_changed`: When the user changes their discover region
- `discover_collection_link_tapped`: When the user taps the external link on a discover collection view
    - A discover collection view is one of the network highlight or guest list
- `discover_show_all_tapped`: When the user taps Show All on a non-curated section this currently includes the trending and popular lists
    - This is a separate event from `discover_list_show_all_tapped`

## To test

```
Prerequisites: Install `debugProd` build variant for testing
```

<details>
    <summary>discover_list_impression</summary>


1.  Launch the app
5. Go to Discover
6. ✅ Verify you see `🔵 Tracked: discover_list_impression ["list_id": "LIST_ID"]` AND `Analytic event discover_list_impression ["list_id": "LIST_ID"]` in Console
7. Verify the `list_id`s are the same in both the tracks
8. Scroll the list
9. ✅ Verify you see `discover_list_impression` events fired for the shown list items (`this behavior is different from iOS where the events are fired at once and not when list items are bound or shown. I kept the previous behavior that we had for Firebase track.`)
</details>

<details>
    <summary>discover_list_show_all_tapped and discover_collection_link_tapped</summary>


1. Go to Discover
2. Locate the Network Highlight
3. Tap on it
4. ✅ Verify you see `🔵 Tracked: discover_list_show_all_tapped ["list_id": "LIST_ID"]` and `Analytic event discover_list_show_all ["list_id": "LIST_ID"]`
   - Where LIST_ID is the list id for the featured item
5. Locate the link row
6. Tap on it
7. ✅ `🔵 Tracked: discover_collection_link_tapped ["list_id": "LIST_ID"]`
   - Where LIST_ID is the same list id for the feature item
8. Go back to Discover
9. Locate the Guest List item
10. Tap on it
11. ✅ Verify you see the `discover_list_show_all_tapped` event with the updated list_id
12. Tap on the link row again
13. ✅ Verify you see the `discover_collection_link_tapped` event with the correct list_id
</details>

<details>
    <summary>discover_list_podcast_tapped and discover_list_podcast_subscribed</summary>


1.  Go to Discover
2. Locate a sponsored podcast (at time of writing one of them is What Could Go Right?)
3. Tap on it
4. ✅ Verify you see: `Analytic event discover_list_podcast_tap` and `🔵 Tracked: discover_list_podcast_tapped` whose values are both: `["podcast_uuid": "PODCAST_UUID", "list_id": "LIST_UUID"]`
   - PODCAST_UUID is the uuid of the podcast that is being promoted
   - LIST_UUID is the uuid of the list that the podcast is featured in
- Tap the Subscribe button
- ✅ Verify you see `Analytic event discover_list_podcast_subscribe and `🔵 Tracked: discover_list_podcast_subscribed` whose values are both: `["list_id": "LIST_ID", "podcast_uuid": "PODCAST_ID"]`
27. Unsubscribe to the podcast
28. Tap back 
29. Tap the + button to subscribe to the podcast from the discover tap (`this behavior was missing for discover_list_podcast_subscribe, I added for it.`)
30. ✅ Verify you see the same logs and values as above
</details>

<details>
    <summary>discover_list_episode_tapped and discover_list_episode_play</summary>

Note: We don't currently have a featured episode live. I added `discover_list_episode_tapped` corresponding to firebase event `discover_list_podcast_episode_tap`.

1. Go to Discover
2. Locate a featured episode (an episode cell with a play button on it)
3. Tap on it
4. ✅ Verify you see: `Analytic event discover_list_podcast_episode_tap` and `🔵 Tracked: discover_list_episode_tapped` whose values are both: `["podcast_uuid": "PODCAST_UUID", "episode_uuid": "EPISODE_UUID", "list_id": "LIST_UUID"]`
   - PODCAST_UUID is the UUID of the podcast the episode is for
   - EPISODE_UUID is the UUID of the episode that was tapped
   - LIST_UUID is the UUID of the list the episode is featured in
5. Tap Back
6.  Tap on the play button
7.   ✅ Verify you see: `Analytic event discover_list_episode_play` and `🔵 Tracked: discover_list_episode_play` whose values are both `"podcast_uuid": "PODCAST_UUID", "episode_uuid": "EPISODE_UUID", "list_id": "LIST_UUID"]`
</details>

<details>
    <summary>discover_show_all_tapped</summary>


1.  Go to Discover
2.  Locate the Trending section
3.  Tap the Show All button
4.  ✅ Verify you see `🔵 Tracked: discover_show_all_tapped ["list_id": "trending"]` in console
5.  Swipe down to the Popular section
6.  Tap the Show All button
7.  ✅ Verify you see `🔵 Tracked: discover_show_all_tapped ["list_id": "popular"]`  in console
</details>

<details>
    <summary>discover_featured_podcast_tapped and discover_featured_podcast_subscribed</summary>


1. Launch the app
2. Go to Discover
3. Locate the featured section at the top of the list
4. Tap on a podcast there
5. ✅ Verify you see `🔵 Tracked: discover_featured_podcast_tapped` in console
6. Tap on the Subscribe `+` button
7. ✅ Verify you see `🔵 Tracked: discover_featured_podcast_subscribed ["podcast_uuid": "PODCAST_UUID"]` - where podcast_uuid is the UUID of the podcast that is being subscribed to

Tracks event `discover_featured_podcast_tapped` maps to Firebase event `featured_podcast_clicked`
Tracks event `discover_featured_podcast_subscribed` maps to Firebase event  `featured_podcast_subscribed`
Firebase events do not include `podcast_uuid` property.

</details>

<details>
    <summary>discover_featured_page_changed</summary>


1. Launch the app
2. Tap on Discover
3. Swipe to the next page in the top featured podcast section
4. ✅  Verify you see `🔵 Tracked: discover_featured_page_changed ["current_page": 1, "total_pages": 5]`
5. Swipe back to the first page
6. ✅ Verify you see `🔵 Tracked: discover_featured_page_changed ["current_page": 0, "total_pages": 5]`
7. Swipe to the last page
9. ✅ Verify you see: `🔵 Tracked: discover_featured_page_changed ["total_pages": 5, "current_page": 4]`

</details>

<details>
    <summary>discover_small_list_page_changed</summary>


1.  Go to Discover
12. Locate the Trending section
13. Swipe to the next page
14. ✅ Verify you see: `🔵 Tracked: discover_small_list_page_changed ["list_id": "trending", "current_page": 1, "total_pages": 5]`
15. ✅ Verify you see `🔵 Tracked: discover_small_list_page_changed ["list_id": "trending", "current_page": 0, "total_pages": 5]`
16. Swipe to the last page
17. ✅ Verify you see: `🔵 Tracked: discover_small_list_page_changed ["list_id": "trending", "total_pages": 4, "current_page": 5]`
38. Swipe down to the Popular section
39. Follow the steps above and verify the `list_id` is `popular`
</details>

<details>
    <summary>discover_region_changed</summary>


1.  Go to Discover
2. Swipe down to the bottom until you see the Select Content Region section
3. Tap on the country button
5. Tap back
6. ✅ Verify you don't see a `discover_region_changed` event in the logs
7. Select a different country
11. Tap back
15. ✅ Verify you see `🔵 Tracked: discover_region_changed ["region": "REGION_CODE"]` - where REGION_CODE is the 2 letter code for the country you selected.
</details>

mchowning reviewed (approved):
I had one question about how we're handling it when the `listUuid` is null, but other than that this looks good.

mchowning commented on modules/features/discover/src/main/java/au/com/shiftyjelly/pocketcasts/discover/view/DiscoverAdapter.kt:
```diff
@@ -371,12 +387,13 @@ internal class DiscoverAdapter(
                             btnSubscribe.setOnClickListener {
                                 btnSubscribe.updateSubscribeButtonIcon(true)
                                 listener.onPodcastSubscribe(podcast = podcast, listUuid = row.listUuid)
+                                row.listUuid?.let { listUuid -> trackDiscoverListPodcastSubscribed(listUuid, podcast.uuid) }
```
Nice catch here. This fixes a bug where we weren't sending this event if the `+` was clicked on a sponsored podcast.

ashiagr commented on modules/features/discover/src/main/java/au/com/shiftyjelly/pocketcasts/discover/view/DiscoverAdapter.kt:
```diff
@@ -201,6 +210,13 @@ internal class DiscoverAdapter(
                 binding.pageIndicatorView.position = position
+                val row = getItem(position) as? DiscoverRow
+                row?.listUuid?.let {
+                    analyticsTracker.track(
```
Trending and popular lists might not have a `listUuid` as I don't see them as nested lists in the admin area. 
I realized a few problems here:

- It should have been best to use the inferred id to track this event as the lists don't have `listUuid` set on them.

- I was doing wrong by passing `position` instead of the `bindingAdapterPosition` to get the `DiscoverRow` id. Here, `position` corresponds to the paginator position and `bindingAdapterPosition` is the position of the `DiscoverRow` w.r.t. the main vertical list. I fixed it in d8841e9. 

I verified that the events were fired correctly now, also matched with iOS:

```
🔵 Tracked: discover_small_list_page_changed, Properties: {"current_page":1,"total_pages":5,"list_id":"trending"
🔵 Tracked: discover_small_list_page_changed, Properties: {"current_page":1,"total_pages":5,"list_id":"popular"
```

ashiagr commented on modules/features/discover/src/main/java/au/com/shiftyjelly/pocketcasts/discover/view/PodcastGridListFragment.kt:
```diff
@@ -164,6 +181,7 @@ open class PodcastGridListFragment : BaseFragment(), Toolbar.OnMenuItemClickList
             linkView.visibility = View.VISIBLE
             linkTextView.text = linkTitle
             linkView.setOnClickListener {
+                analyticsTracker.track(AnalyticsEvent.DISCOVER_COLLECTION_LINK_TAPPED, mapOf(LIST_ID_KEY to (listUuid ?: NONE)))
```
>  I like how you're using this "none" value here if the listUuid is missing. Curious why you decided to add this here.

`listUuid` passed to `PodcastGridListFragment` and `PodcastListFragment` can be null. It would have been best to pass `inferredId` property in this case. I made changes in `75fdef8` to get `inferredId` from the arguments passed to the fragment. 

I also made sure that I used `inferredId` at places [matching those with iOS](https://github.com/Automattic/pocket-casts-ios/search?q=inferredListId).

> In most other places we are instead just not sending the event at all if the listUuid is null (doing a text search for listUuid?.let seems to pull up a lot of examples. 

I ran through other places and found that `listUuid` is always present for them. The only exceptions I found (from admin area) are:
- `popular` and `trending` types lists which do not have `listUuid` and 
- `featured` type podcasts that have podcast uuid set on them. 

Moreover, `listUuid?.let` check was already present at most places for Firebase events. I didn't want to track additional events with `listUuid` set to `none`. Let me know if you feel otherwise. 🤯  

> I actually ran into an instance where an event wasn't getting fired because the listUuid was null. See https://github.com/Automattic/pocket-casts-android/pull/345#discussion_r986122711.

I answered it [here](https://github.com/Automattic/pocket-casts-android/pull/345#discussion_r986489681).

ashiagr commented:
Thanks for reviewing the PR, @mchowning! I've addressed your comments, I'll go ahead and merge the PR before the code freeze. Feel free to report any issues and I'll handle them separately.
